src/base/bios/bios.S

```diff
@@ -875,6 +875,24 @@ MSDOS_lw_entry_cs:
 .macro int_rvc inum
 	.globl int_rvc_start_\inum
 int_rvc_start_\inum:
+	jmp 31f		// EB xx
+int_rvc_data_\inum:
+
+	.globl int_rvc_ip_\inum
+int_rvc_ip_\inum:
+	.word 0
+
+	.globl int_rvc_cs_\inum
+int_rvc_cs_\inum:
+	.word 0
+
+	.word 0x424B	// signature "KB"
+	.byte 0		// flag
+	jmp 30f		// EB xx for compat
+	.space 7	// padding
+30:
+	lret
```

Comment:
Please comment this lret, I am puzzled at it.

Reply:
Yes it's in the spec. Apparently the sharing protocol was originally devised for hardware, and it's a leftover from that. Here's the section from the spec, the last paragraph is relevant:
~~~
The block looks like this:

    intercept:  jmp short int_handler
    prevhndlr   dd 0
    signature   dw 424Bh
    flag        db 0
                jmp short hwreset
                db 7 dup (0)            ; Reserved
    int_handler:
                ; your interrupt handler starts here...

'intercept' is the address you'll use when you do a SETVEC to intercept
the interrupt vector, i.e., the entry point for your interrupt handler.

'prevhndlr' is set to contain the initial contents of the interrupt
vector at the time you take over.  (This is what we were calling
OldInt21 in the previous sections.)

'signature' must contain 424Bh ("KB") and is used to help identify one
of these blocks.

'flag' is important only if you're taking over a hardware interrupt
that requires an EOI (say, INT 8 or INT 9).  For software interrupts
(INT 16 or INT 21, e.g.), leave it 0. For hardware interrupts, the
first installed handler should set the flag to 80h.  Only the handler
whose flag is 80h is allowed to issue EOI.

The 'jmp short hwreset' is pretty much irrelevant to anything software
people will use; it's primarily for hardware manufacturers (allows them
to specify code to be executed to reinitialize the hardware on a system
reset).  However, you must be prepared for the eventuality that this
entry point will be called by someone; do this by simply defining an
HWRESET label with a RETF:

    hwreset:    retf

Finally, leave the seven reserved bytes as zeroes.
~~~

Reply:
So I think you did it wrongly.
It says "jmp short", which is a 2-byte insn EB xx in this context
(can be larger), and retf is 1-byte. The struct will be mis-aligned.
I think you need to put "lret" somewhere below and jump to it.

Reply:
~~~
dosdebug> d int_rvc_start_33
dosdebug> 

f000:f600 EB 11 8A 11 D8 00 4B 42 00 EB 07 00 00 00 00 00  k...X.KB.k......
f000:f610 00 00 CB 66 C1 E0 10 66 C1 E3 10 B0 42 B3 01 B4  ..KfA`.fAc.0B3.4
~~~

I did swap the jmp targets around, but I think that's outside the significant part of the header. Here's what I have from the dump above:
```asm
EB 11 # jmp short to real code
8A 11 D8 00 # long prev addr
42 4B  # signature  ("KB")
00 #  flag
EB 07 # jmp short to hwreset
00 00 00 00 00 00 00 # seven bytes padding
```
```asm
CB (hwreset) # retf
<.... real handler code ....>
``` 

The location of the label hwreset in their example could be anywhere as it follows the handler code. I swapped it (label 30:) to be before the handler code, this ensures it's a `jmp short` to reach it. I guess it comes down to whether we think the start address of the handler is significant?

I'll swap it back.

Reply:
Ah, indeed, there IS a jmp, and the lret is
outside of the header. You said "Yes it's in the spec."
and I took it like in the header too.